### ISSUE TITLE
Add known Lexis Nexis endpoints to be blocked 

### ISSUE THREAD

**Issue by ACK-J** [mullvad/dns-adblock#6]
# Tl;dr
I spent a lot of time researching a script prevalent on the internet that invades user's privacy and bypasses all security measures such as antivirus and ad-blockers. I have provided a list of domains that are used to pull the script down [HERE](https://gist.github.com/ACK-J/65dfe84fcf5a06c46364e5f2bd29c118)

# Why?
- Lexis Nexis is one of the largest data brokers in the world and bought a company called "Threat Metrix" which produces scripts to heavily fingerprint users and try to determine their "True Location" going as far as port scanning a user’s network with javascript web-sockets. These scripts are stealthily pulled down from a customer-specific subdomain owned by the website you are visiting but have a CNAME record that redirects to `*.online-metrix.net` to pull down these scripts, making them very difficult to block. 

# How it works
![image](https://user-images.githubusercontent.com/60232273/122772128-0f9ebd00-d275-11eb-8a88-9a8af156ab26.png)

# How invasive is the script?
- The data being exfiled is encrypted into an image with XOR.
- The javascript is assembled via string.join (like malware often does) and then executed in a service worker.
- Each time you load the page, the javascript is re-obfuscated.
- It collects 416 data points about your computer / network. Shown [HERE](https://gist.github.com/ACK-J/aa8dceb072d31d97a4e7fe0ef389f370)
- Port scans your computer
- Trys to evade being blocked by having their customers set up random subdomains, with no standard scheme, to redirect to their backend servers
- Data can be used to fingerprint users and track them even behind a VPN or just to sell their data

# Solution
Luckily this annoyed me enough to find multiple solutions to the problem. I wrote a python script that uses shodan to locate a majority of these customer-specific endpoints. The script can be found [HERE](https://gist.github.com/ACK-J/7a2da401c732cbe58479d03acc4e4b43) I also wrote a FOSS firefox extension to block port scanning in general and dynamically resolve the CNAME's and block any that go to threat metrix infrastructure, but that is beside the point [HERE](https://addons.mozilla.org/en-US/firefox/addon/port-authority/)

- I have run this script collecting new endpoints for a few months and have published all of them [HERE](https://gist.github.com/ACK-J/65dfe84fcf5a06c46364e5f2bd29c118) My request is to please block every domain in this list! :)
- This list will never 100% include every endpoint, just the ones that I can find. There are about ~450 endpoints in there so I am pretty sure it covers a majority. The only way to ensure 100% that your computer never pulls down one of these scripts is to install the firefox addon I wrote since it resolves the CNAME and determines it that way. 

# How to verify there is no False Positives

`while read line; do if [ "$line" != "" ]; then dig @1.1.1.1 +short "$line" cname >> out.txt; fi done < <(cat threatmetrix.txt)`

- If you run the bash one-liner above, make sure you pass the right file name to STD-IN,  it will use `dig` to grab the CNAME for each customer specific endpoint I provided and they all should be in the following format `h-<company name>.online-metrix.net.`

# References
- Here is a great post from DJ Nemec who reverse-engineered this malware [HERE](https://blog.nem.ec/2020/05/24/ebay-port-scanning/)
- If you remember in May of 2020 Ebay got caught port scanning their users, what actually was happening was they hired this company to run the script pulled down by these endpoints. [HERE](https://www.ghacks.net/2020/05/25/ebay-is-port-scanning-your-system-when-you-load-the-webpage/)

**Comment by jbjorkang** [comment]
Thanks a lot for your hard work here! Very much appreciated. We will keep this in mind for future updates to the block lists!
